chain/bitcoind_conn.go

```diff
@@ -471,6 +607,104 @@ func (c *BitcoindConn) txEventHandler() {
 	}
 }
 
+// txEventHandlerRPC is a goroutine that uses the RPC client to check the
+// mempool for new transactions.
+func (c *BitcoindConn) txEventHandlerRPC() {
+	defer c.wg.Done()
+
+	log.Info("Started polling for new bitcoind transactions via RPC.")
+	ticker := time.NewTicker(c.pollTxTimer)
+
+	for {
+		select {
+		case <-ticker.C:
+			// Every few seconds, we poll the mempool to check for
+			// transactions we haven't seen yet.
+			txs, err := c.client.GetRawMempool()
+			if err != nil {
+				log.Errorf("Unable to retrieve mempool txs: "+
+					"%v", err)
+			}
+
+			mempool := make(map[chainhash.Hash]struct{})
+
+			c.rescanClientsMtx.Lock()
+			if c.rescanClients == nil || len(c.rescanClients) < 1 {
+				log.Errorf("No client added yet")
+				continue
+			}
+
+			// Grab one of the clients from the rescanClients map
+			// so we can look at its mempool.
+			var client *BitcoindClient
+			idx := 0
+			for _, currClient := range c.rescanClients {
+				if idx > 0 {
+					break
+				}
+
+				client = currClient
+
+				idx += 1
+			}
+			c.rescanClientsMtx.Unlock()
+
+			for tx, _ := range client.mempool {
+				client.watchMtx.Lock()
+				mempool[tx] = struct{}{}
+				client.watchMtx.Unlock()
+			}
+
+			// We'll scan through the most recent txs in the
+			// mempool to see whether there are new txs that we
+			// need to send to the client.
+			for _, txHash := range txs {
+				// If the transaction isn't in the local
+				// mempool, we'll send it to all of the
+				// clients.
+				if _, ok := mempool[*txHash]; !ok {
+					// Grab full mempool transaction from hash.
+					tx, err := c.client.GetRawTransaction(
+						txHash,
```

Comment:
So every 3 seconds, this is getting the raw mempool and then also calling `GetRawTransaction` on every transaction in the mempool? Just trying to gauge load on the bitcoin node and how many requests we're doing.

Reply:
Every X seconds it gets and looks at the raw mempool. But it should only call getrawtransaction when it finds a tx we haven't seen already. Each client keeps a mempool map with all of the transactions that have already been looked at, so we check if it's in there or not yet on line #665

Though actually now that I think about it... the mempool map that exists rn in each client only holds txs that have been "filtered" -- which in the instance of an lnd node, I think are txs that open/close channels(?). So I'm thinking of adding another fullMempool map that doesn't apply the filter, to decrease the amount of times we call getrawtransaction. 

Is that kinda what you meant? Thoughts?

Reply:
Yeah that makes sense. Do you think we need a map of unfiltered txs? I think if there's a tx in the mempool that isn't for a channel open/close we wouldn't care about it anyway, right?

Additionally, do transactions leave this mempool map? Like after a certain amount of time or if they get confirmed? Just want to make sure we don't have some sort of memory leak where we're constantly adding txs but never removing them.

Reply:
Yeah I was thinking the same thing, but the problem is we don't filter the tx until after receiving the full transaction here, so I think we need the full mempool map or something along those lines if we want to avoid using GetRawTransaction for txs we've already seen

And yes, lucky for us there is a mechanism for removing txs from the mempool map already in place! When a tx leaves the mempool it is placed in the "expiredMempool" map. Then any transactions that are 288 blocks or older are deleted entirely. So I'm thinking we can just reuse that process for the fullMempool map as well

Soundin' okay? And if so, I'll add some of these decisions to the comments so it's more clear

Reply:
Ok yeah this sounds good. Correct me if I'm wrong but this sounds like a tradeoff between using more memory or making more requests to the bitcoin node. Obviously I want both but that's not how this works haha 😁 . So we'd have to create an additional map of the mempool to prevent hitting the bitcoin node a lot. I wonder if it'd be ok to update the existing map to not have a filter? So instead of 2 maps, we have 1 that's unfiltered? All in all, I think we definitely should be optimizing for least amount of load on the Bitcoin node. However you can make that happen is cool with me.

Sounds good on the expiration too

Reply:
Got it! Yes that sounds exactly right, that's the tradeoff. I thought about this some more and I see three options:

A) Just doing some rough calculations to get a better outlook on the fullMempool map... it holds tx hashes, which are 32 bytes. So when the mempool has a ton of transactions, based on historical data this could grow to as large as 4 MB (though usually its much less than this). The fullExpiredMempool map would be quite a bit bigger... something like between 20 to 30 MB based on how many transactions are in each block

B) Every three seconds we make like roughly 5,000-10,000 (though can go as high as 200,000) requests to GetRawTransaction, which seems like a ton? 

C) Using a mix of these two things. Use the fullMempool map (with an upper bound of 4 MB), but ditch the fullExpiredMempool map (20MB to 30MB). Instead of the fullExpiredMempool map, every time we receive a new block we just wipe the fullMempool map and start over. The disadvantage is that every time we wipe it we need to make a ton of requests to refill the mempool (depending on how many txs are in the mempool). But it's only once every ~10 minutes when a new block comes in, vs every 3 seconds in scenario B.

Sorry for the wall of text LOL. My thinking is to go with C? But def. want to see what you think, since much more aware to how much total memory/cpu/etc resources a Voltage Lightning node eats up and how this adds to it. Thoughts?

Reply:
Ohhh I'm looking into the 288 number and it looks like it's around for reorgs: https://github.com/btcsuite/btcwallet/blob/fe61cc5d7886913ae605f7f1912fc4f21120a109/chain/bitcoind_client.go#L79. So if I'm understanding right, that component is needed to help keep track historical filtered txs. 

But you're right, for our purposes, it doesn't seem necessary to keep 288 blocks... As I think about it, I think we could have a fullExpiredMempool map, but instead of keeping track of which block a transaction makes it in, keep track of the hour it was introduced. Then have a goroutine that runs every few hours (or another time frame) and prunes anything older than a few hours that might be an issue. That way the fullExpiredMempool map can be limited to just holding tx from a few hours. So that would be like... 1-2MB instead of the 20MB estimated above.

The reason I suggest fullExpiredMempool holding hours a tx is introduced instead of blocks... the original memory leak possibility that you had pointed out. The original mechanism in the btcwallet code is to delete transactions from the mempool map as they're getting into blocks. But I think there's still a possibility that a tx gets into the mempool map but just never gets into a block? Not really sure how common of a scenario that is, but... I"m assuming when bitcoin's mempool fills up, this could be an issue?

Not sure if that sounds sane. Does that sound sane? Lol

And yes, I can totally add a flag if we still want to enable a couple of different options !

Reply:
Ah, that's interesting regarding the 288 blocks... Do you think we should maybe keep it as is given it's for perhaps an important reason? While I definitely want this to fit our use case, I want this to be able to go into the upstream project and not a fork we maintain. So I don't think your idea is a bad one, I just want to make sure it's acceptable to the upstream project. Transactions can definitely enter the mempool and never get confirmed. I'm not sure how btcwallet handles those cases. 

https://bitcoin.stackexchange.com/questions/21901/what-happens-if-your-transaction-is-never-confirmed

I may be overcomplicating this so I think I should defer to your judgement. I'm totally ok with your solution you suggested above, I just want to make sure it's compliant to the project. I think we should have this default to the high memory option with a flag you can set to use less memory, but call the bitcoin node more frequently. like we talked about 2 posts ago. Does this make sense? Whatever we can do to optimize around memory consumption and mempool management is cool with me. Thanks for doing this!

Reply:
Ok cool. Thanks for talking this through with me, it's been super helpful! The more I think about it, the more I'm thinking the last option in the last post makes the most sense tbh. But I think I'll run it by Oli to make sure it's a sensible approach (or if there's someone who you think might be better to reach out to lmk)

Reply:
Yeah Oli would be a good person for this. Can you elaborate on which method you think is best? This has been really helpful to think through this and decide what the tradeoffs are.

Reply:
Sure thing! Yeah so I'm thinking when we started this conversation that I was totally misunderstanding the mechanism that was already in place in btcwallet for keeping a mempool map. I'm now thinking the 288 block thing is something we don't need to replicate for our purposes -- it's there to shortcut grabbing transactions that were in the mempool -- and then are confirmed, which we don't need because we're just trying to find new txs that pop into the mempool, not txs that have been confirmed

I'll confirm with Oli! It seems complicated so def. want to make sure I'm understanding. But If it's the case, I think we just want to use a mempool map (which could grow as big as 4MB) and we won't really need a fullExpiredMempool (freeing up 20-30MB of memory)

BUT, to make things super confusing (lol), the reason I think we *might* still want a smaller fullExpiredMempool map, as I mentioned here https://github.com/orbitalturtle/btcwallet/pull/1#discussion_r770239926 is that... from what I can tell of the mempool map mechanism that's in place for btcwallet, it only deletes from the mempool map when a tx is confirmed. It doesn't ever delete txs that never confirm, which seems like a possible memory leak, but I could be missing something. So, this might be a problem already existing in btcwallet. If it is, I want to make sure we at least don't duplicate it

Anyway will talk to Oli to see if what I'm thinking makes any sense and I'll get back to you!

Reply:
This makes sense.  Thanks! yeah I'd be curious about that edge case of transactions that never confirm. Something to definitely think about. Oli may have good ideas.

One possible idea I had was to add a timestamp of when we saw the transaction. Then have some go routine that runs once a day or something and removes any unconfirmed transaction that's older than X days. Might be better options but just throwing it out there.

Reply:
Ok cool. Confirmed with Oli that this all makes sense. Re: the goroutine that runs once a day, Oli said the same thing as you! I'll start coding this up and should get all the changes done here and from the rest of your review done by the end of the week :-)